File: src/main/java/com/github/sgtsilvio/gradle/android/retrofix/transform/MethodMap.java
package com.github.sgtsilvio.gradle.android.retrofix.transform;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.HashMap;

/**
 * @author Silvio Giebl
 */
public class MethodMap extends HashMap<String, MethodMap.Entry> {

    private void put(final @NotNull String methodSignature, final @NotNull String type, final @NotNull String replacement) {
        put(methodSignature, new Entry(type, replacement, get(methodSignature)));
    }

    public @NotNull ForType forType(final @NotNull String type) {
        return new ForType(type);
    }

    public static class Entry {
        public final @NotNull String type;
        public final @NotNull String replacement;
        public final @Nullable Entry next;

        Entry(final @NotNull String type, final @NotNull String replacement, final @Nullable Entry next) {
            this.type = type;
            this.replacement = replacement;
            this.next = next;
        }
    }

    public class ForType {
        final @NotNull String type;

        ForType(final @NotNull String type) {
            this.type = type;
        }

        public @NotNull ForType redirect(
                final @NotNull String method, final @NotNull String signature, final @NotNull String replacement) {

            return redirect(method, signature, replacement, false);
        }

        public @NotNull ForType redirectStatic(
                final @NotNull String method, final @NotNull String signature, final @NotNull String replacement) {

            return redirect(method, signature, replacement, true);
        }

        private @NotNull ForType redirect(
                final @NotNull String method, final @NotNull String signature, @NotNull String replacement,
                final boolean isStatic) {

            if (!signature.endsWith("V")) {
                replacement = "$_ = " + replacement;
            }
            put(method + " " + signature, type, replacement + "." + method + "(" + (isStatic ? "" : "$0,") + "$$);");
            return this;
        }
    }
}

File: src/main/kotlin/com/github/sgtsilvio/gradle/android/retrofix/transform/MethodMap.kt
package com.github.sgtsilvio.gradle.android.retrofix.transform

/**
 * @author Silvio Giebl
 */
class MethodMap {

    private val map = HashMap<String, Entry>()

    fun get(name: String, descriptor: String) = map["$name$descriptor"]

    private fun redirect(
        owner: String,
        isStatic: Boolean,
        name: String,
        descriptor: String,
        newOwner: String,
        newName: String,
    ) {
        val key = "$name$descriptor"
        map[key] = Entry(owner, isStatic, name, descriptor, newOwner, newName, map[key])
    }

    fun forType(type: String) = ForType(type)

    class Entry(
        val owner: String,
        val isStatic: Boolean,
        val name: String,
        val descriptor: String,

        val newOwner: String,
        // no newIsStatic as always static
        val newName: String,
        // newDescriptor is computed

        val next: Entry?,
    ) {
        val newDescriptor get() = if (isStatic) descriptor else "(L" + owner + ";" + descriptor.substring(1)
    }

    inner class ForType(private val owner: String) {

        fun redirect(name: String, descriptor: String, newOwner: String): ForType {
            redirect(owner, false, name, descriptor, newOwner, name)
            return this
        }

        fun redirectStatic(name: String, descriptor: String, newOwner: String): ForType {
            redirect(owner, true, name, descriptor, newOwner, name)
            return this
        }

        fun redirectStatic(method: String, descriptor: String, newOwner: String, newName: String): ForType {
            redirect(owner, true, method, descriptor, newOwner, newName)
            return this
        }
    }
}
